File: Grammars/LUA/LUAGrammar.vb
Imports NLua
Imports System.Xml.Serialization
Imports System.Text
Imports System.Windows
Imports System.IO

Namespace LUA
    Public Class LUAGrammar
        Inherits BaseGrammar

        Private _script As String
        Private _scriptPath As String
        Private _output As StringBuilder

        <XmlElement("script")>
        Public Property Script As String
            Get
                Return _script
            End Get
            Set(value As String)
                If value <> _script Then
                    _script = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        <XmlElement("scriptPath")>
        Public Property ScriptPath As String
            Get
                If Path.IsPathRooted(_scriptPath) Then Return _scriptPath
                Return Path.Combine(Path.GetDirectoryName(Reflection.Assembly.GetCallingAssembly.Location), "DataFiles", _scriptPath)
            End Get
            Set(value As String)
                _scriptPath = value
            End Set
        End Property

        Default Public Overrides ReadOnly Property Item(columnName As String) As String
            Get
                If columnName = "Script" AndAlso String.IsNullOrWhiteSpace(Script) Then Return "Script is required."
                Return MyBase.Item(columnName)
            End Get
        End Property

        Public Overrides Function Analyze() As String
            Return "Not Supported"
        End Function

        Protected Overrides Function GenerateName() As String
            Dim l As New NLua.Lua
            Dim name As New StringBuilder
            Dim calc As New Calculator(Me, l)
            Dim modulePath As String = String.Empty

            Try
                _output = Nothing
                For Each path As String In Utility.GrammarFilePaths
                    modulePath &= ";" & IO.Path.Combine(path, "lua")
                    modulePath &= "\?.lua"
                Next

                modulePath = modulePath.Replace("\", "\\")

                For Each parameter As Parameter In Parameters
                    l(parameter.Name) = parameter.Value
                Next

                l("Grammar") = Me
                l("Me") = l
                l.RegisterFunction("print", Me, Me.GetType.GetMethod("Print"))
                l.RegisterFunction("printif", Me, Me.GetType.GetMethod("PrintIf"))
                l.DoString("luanet.load_assembly('Grammars')")
                l.DoString("CalcClass=luanet.import_type('Grammars.LUA.Calculator')")
                l.DoString("calc=CalcClass(Grammar,Me)")
                l.DoString("package.path = package.path .. '" & modulePath & "'")
                l.DoString("import = function() end")

                If String.IsNullOrWhiteSpace(Script) Then
                    l.DoFile(ScriptPath)
                Else
                    l.DoString(Script)
                End If

                If _output Is Nothing Then Return String.Empty
                Return _output.ToString()
            Catch ex As NLua.Exceptions.LuaScriptException
                If _output IsNot Nothing Then
                    Return _output.ToString & "<br /><br /><span style='color:red;'>" & ex.ToString & "</span>"
                Else
                    Return "<span style='color:red;'>" & ex.ToString & "</span>"
                End If

            End Try
        End Function

        Public Sub Print(ByVal value As Object)
            If _output Is Nothing Then _output = New StringBuilder
            _output.AppendLine(value.ToString)
        End Sub

        Public Sub PrintIf(ByVal condition As Boolean,
                           ByVal value As Object)
            If condition Then
                Print(value)
            End If
        End Sub
    End Class
End Namespace

File: Grammars/LUA/LUAGrammar.cs
using System.IO;
using System.Reflection;
using System.Text;
using System.Xml.Serialization;
using NLua;
using NLua.Exceptions;

namespace Grammars.LUA
{
    public class LUAGrammar : BaseGrammar
    {
        private StringBuilder _output;

        private string _script;
        private string _scriptPath;

        [XmlElement("script")]
        public string Script
        {
            get { return _script; }
            set
            {
                if (value != _script)
                {
                    _script = value;
                    OnPropertyChanged();
                }
            }
        }

        [XmlElement("scriptPath")]
        public string ScriptPath
        {
            get
            {
                if (Path.IsPathRooted(_scriptPath))
                {
                    return _scriptPath;
                }
                return Path.Combine(
                    Path.GetDirectoryName(Assembly.GetCallingAssembly().Location),
                    "DataFiles",
                    _scriptPath);
            }
            set { _scriptPath = value; }
        }

        public override string this[string columnName]
        {
            get
            {
                if (columnName == "Script" && string.IsNullOrWhiteSpace(Script))
                {
                    return "Script is required.";
                }
                return base[columnName];
            }
        }

        public override string Analyze()
        {
            return "Not Supported";
        }

        protected override string GenerateName()
        {
            var l = new Lua();
            var modulePath = string.Empty;

            try
            {
                _output = null;
                foreach (var path in Utility.GrammarFilePaths)
                {
                    modulePath += ";" + Path.Combine(path, "lua");
                    modulePath += "\\?.lua";
                }

                modulePath = modulePath.Replace("\\", "\\\\");

                foreach (var parameter in Parameters)
                {
                    l[parameter.Name] = parameter.Value;
                }

                l["Grammar"] = this;
                l["Me"] = l;
                l.RegisterFunction("print", this, GetType().GetMethod("Print"));
                l.RegisterFunction("printif", this, GetType().GetMethod("PrintIf"));
                l.DoString("luanet.load_assembly(\'Grammars\')");
                l.DoString("CalcClass=luanet.import_type(\'Grammars.LUA.Calculator\')");
                l.DoString("calc=CalcClass(Grammar,Me)");
                l.DoString("package.path = package.path .. \'" + modulePath + "\'");
                l.DoString("import = function() end");

                if (string.IsNullOrWhiteSpace(Script))
                {
                    l.DoFile(ScriptPath);
                }
                else
                {
                    l.DoString(Script);
                }

                if (_output == null)
                {
                    return string.Empty;
                }
                return _output.ToString();
            }
            catch (LuaScriptException ex)
            {
                if (_output != null)
                {
                    return _output + "<br /><br /><span style=\'color:red;\'>" + ex + "</span>";
                }
                return "<span style=\'color:red;\'>" + ex + "</span>";
            }
        }

        public void Print(object value)
        {
            if (_output == null)
            {
                _output = new StringBuilder();
            }
            _output.AppendLine(value.ToString());
        }

        public void PrintIf(bool condition, object value)
        {
            if (condition)
            {
                Print(value);
            }
        }
    }
}